inference-engine/thirdparty/clDNN/kernel_selector/core/actual_kernels/non_max_suppression/non_max_suppression_kernel_ref.cpp

```diff
@@ -0,0 +1,304 @@
+// Copyright (C) 2021 Intel Corporation
+// SPDX-License-Identifier: Apache-2.0
+//
+
+#include "non_max_suppression_kernel_ref.h"
+#include "kernel_selector_utils.h"
+#include <vector>
+
+namespace kernel_selector {
+
+Datatype NonMaxSuppressionKernelRef::GetAccumulatorType(const non_max_suppression_params& params) const {
+    auto in_dt = params.inputs[0].GetDType();
+    auto out_dt = params.output.GetDType();
+
+    auto smaller_fp_type = [](const Datatype& current, const Datatype& candidate) -> Datatype {
+        if (candidate != Datatype::F32 || candidate != Datatype::F16)
+            return current;
+
+        return BytesPerElement(candidate) < BytesPerElement(current) ? candidate : current;
+    };
+
+    Datatype fp_type = Datatype::F32;
+    fp_type = smaller_fp_type(fp_type, in_dt);
+    fp_type = smaller_fp_type(fp_type, out_dt);
+
+    return fp_type;
+}
+
+ParamsKey NonMaxSuppressionKernelRef::GetSupportedKey() const {
+    ParamsKey k;
+    k.EnableInputDataType(Datatype::INT32);
+    k.EnableInputDataType(Datatype::F16);
+    k.EnableInputDataType(Datatype::F32);
+    k.EnableOutputDataType(Datatype::INT32);
+    k.EnableOutputDataType(Datatype::INT64);
+    k.EnableDifferentTypes();
+    k.EnableAllInputLayout();
+    k.EnableAllOutputLayout();
+    k.EnableTensorOffset();
+    k.EnableTensorPitches();
+    k.EnableBatching();
+    return k;
+}
+
+JitConstants NonMaxSuppressionKernelRef::GetJitConstants(const non_max_suppression_params& params) const {
+    JitConstants jit = MakeBaseParamsJitConstants(params);
+
+    const auto& input0 = params.inputs[0];
+    switch (input0.GetDType()) {
+    case Datatype::F32:
+        jit.AddConstant(MakeJitConstant("TO_COORD_TYPE_4", "convert_float4"));
+        jit.AddConstant(MakeJitConstant("COORD_TYPE_4", "float4"));
+        jit.AddConstant(MakeJitConstant("TO_COORD_TYPE", "convert_float"));
+        jit.AddConstant(MakeJitConstant("COORD_TYPE", "float"));
+        break;
+
+    case Datatype::F16:
+        jit.AddConstant(MakeJitConstant("TO_COORD_TYPE_4", "convert_half4"));
+        jit.AddConstant(MakeJitConstant("COORD_TYPE_4", "half4"));
+        jit.AddConstant(MakeJitConstant("TO_COORD_TYPE", "convert_half"));
+        jit.AddConstant(MakeJitConstant("COORD_TYPE", "half"));
+        break;
+
+    default:
+        throw std::invalid_argument("NMS input0 type should be one of F32 or F16.");
+    }
+
+    jit.AddConstants({MakeJitConstant("SORT_RESULT_DESCENDING", params.sort_result_descending),
+                      MakeJitConstant("BOX_ENCODING", static_cast<int>(params.box_encoding))});
+
+    jit.AddConstant(MakeJitConstant("OUTPUT_NUM", params.output.Batch().v));
+
+    if (params.num_select_per_class_type == NmsArgType::Input) {
+        char inputTypeStr[128];
+        snprintf(inputTypeStr, sizeof(inputTypeStr), "INPUT%d_TYPE", params.GetIndexNumSelectPerClass());
+        jit.AddConstant(MakeJitConstant("NUM_SELECT_PER_CLASS_TYPE", std::string(inputTypeStr)));
```

Comment:
I think something as following will be simpler and more readable:
```cpp
auto idx = params.GetIndexNumSelectPerClass();
jit.AddConstant(MakeJitConstant("NUM_SELECT_PER_CLASS_TYPE", toCLType(params.inputs[idx].GetDType())));
```

Maybe some helper would be useful as similar code appears multiple times:
```cpp
std::string GetInputTypeStr(uint32_t idx) {
    return toCLType(params.inputs[idx].GetDType());
}

jit.AddConstant(MakeJitConstant("NUM_SELECT_PER_CLASS_TYPE", GetInputTypeStr(params.GetIndexNumSelectPerClass())));